src/serialization/serialization.rs

```diff
@@ -0,0 +1,412 @@
+                // skip the zeros
+                dest_index = dest_index.checked_add(zero_count)
+                    .ok_or(DeserializeError::UsizeTypeTooSmall)?;
+                continue;
+            } else {
+                let count: T = T::from_i64(num)
+                    .ok_or(DeserializeError::UnsuitableCounterType)?;
+
+                h.set_count_at_index(dest_index, count)
+                    .map_err(|_| DeserializeError::EncodedArrayTooLong)?;
+
+                dest_index = dest_index.checked_add(1)
+                    .ok_or(DeserializeError::UsizeTypeTooSmall)?;
+            }
+        }
+
+        // TODO destat is expensive; should accumulate the necessary state while deserializing
+        // dest_index is one past the last written index, and is therefore the length to scan
+        h.restat(dest_index);
+
+        Ok(h)
+    }
+}
+
+/// Encode counts array into slice.
+/// The slice must be at least 9 * the number of counts that will be encoded.
+fn encode_counts<T: Counter>(h: &Histogram<T>, buf: &mut [u8]) -> Result<usize, SerializeError> {
+    let index_limit = h.index_for(h.max()) + 1;
+    let mut index = 0;
+    let mut bytes_written = 0;
+
+    assert!(index_limit <= h.counts.len());
+
+    while index < index_limit {
+        // index is inside h.counts because of the assert above
+        let count = unsafe { *(h.counts.get_unchecked(index)) };
+        index += 1;
+
+        // Non-negative values are positive counts or 1 zero, negative values are repeated zeros.
+
+        let mut zero_count = 0;
+        if count == T::zero() {
+            zero_count = 1;
+
+            // index is inside h.counts because of the assert above
+            while (index < index_limit) && (unsafe { *(h.counts.get_unchecked(index)) } == T::zero()) {
+                zero_count += 1;
+                index += 1;
+            }
+        }
+
+        let count_or_zeros: i64 = if zero_count > 1 {
+            // zero count can be at most the entire counts array, which is at most 2^24, so will fit.
+            -zero_count
+        } else {
+            count.to_i64().ok_or(SerializeError::CountNotSerializable)?
+        };
+
+        let zz = zig_zag_encode(count_or_zeros);
+
+        bytes_written += varint_write(zz, &mut buf[bytes_written..]);
+    }
+
+    Ok(bytes_written)
+}
+
+/// Write a number as a LEB128-64b9B little endian base 128 varint to buf. This is not
+/// quite the same as Protobuf's LEB128 as it encodes 64 bit values in a max of 9 bytes, not 10.
+/// The first 8 7-bit chunks are encoded normally (up through the first 7 bytes of input). The last
+/// byte is added to the buf as-is. This limits the input to 8 bytes, but that's all we need.
+/// Returns the number of bytes written (in [1, 9]).
+#[inline]
+fn varint_write(input: u64, buf: &mut [u8]) -> usize {
+    // The loop is unrolled because the special case is awkward to express in a loop, and it
```

Comment:
Maybe, but I'm not at all convinced that this is the final form of the algorithm so I'd prefer to wait until it can't get any faster before we make it pretty.